lib/absinthe/phase/document/result.ex

```diff
@@ -79,8 +79,13 @@ defmodule Absinthe.Phase.Document.Result do
     Map.merge(left, right, &do_deep_resolve/3)
   end
 
-  defp do_deep_resolve(_key, [%{} = left], [%{} = right]) do
-    [do_deep_merge(right, left)]
+  defp do_deep_resolve(_key, %{} = left, %{} = right) do
+    do_deep_merge(left, right)
+  end
+
+  defp do_deep_resolve(_key, [_|_] = left, [_|_] = right) when length(left) == length(right) do
```

Comment:
Ah yeah that makes sense, if the resolution functions have some kind of side cause like loading from the DB, and the record happens to change between 2 invocations of the function, you're out of luck.

Quite honestly I think this needs to be solved at the blueprint level if possible. However it may mean enumerating all possible combinations of type conditions on a selection set. While algorithmically obnoxious, it seems like in practice the N of type conditions is wildly smaller than the M of actual results. Need to think about this more.